ocaml/xapi/xapi_host.ml

```diff
@@ -1366,45 +1366,92 @@ let certificate_sync ~__context ~host = Certificates.local_sync ()
 let get_server_certificate ~__context ~host =
   Certificates.get_server_certificate ()
 
+let add_certificate_to_db ~__context ~host certificate
+    root_fingerprints =
+  let date_of_ptime time = Date.of_float (Ptime.to_float_s time) in
+  let dates_of_ptimes (a, b) = (date_of_ptime a, date_of_ptime b) in
+  let not_before, not_after =
+    dates_of_ptimes (X509.Certificate.validity certificate)
+  in
+  let raw_fingerprint =
+    X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) certificate
+  in
+  let fingerprint = Certificates.pp_hash raw_fingerprint in
+  let uuid = Uuid.(to_string (make_uuid ())) in
+  let ref = Ref.make () in
+  let _type =
+    if List.mem raw_fingerprint root_fingerprints then `host_and_pool else `host
+  in
+  Db.Certificate.create ~__context ~ref ~uuid ~_type ~host ~pool:Ref.null
+    ~not_before ~not_after ~fingerprint ;
+  (* Link host to trust anchors *)
+  let root_certs =
+    root_fingerprints
+    |> List.map (fun root_fingerprint ->
+           let expr =
+             Db_filter_types.(
+               Eq
+                 ( Field "fingerprint"
+                 , Literal (Certificates.pp_hash root_fingerprint) ))
+           in
+           Db.Certificate.get_refs_where ~__context ~expr)
+    |> List.flatten
+    |> List.sort_uniq Stdlib.compare
+  in
+  Db.Certificate.set_validated_by ~__context ~self:ref ~value:root_certs ;
+  ref
+
+let uninstall_outdated_certificates ~__context outdated_server_certs =
```

Comment:
This is a misleading name since we are not checking that certs are outdated here ( `let uninstall_certificates ~__context certs` is more accurate ? )